### PR TITLE
prohibit the lhs of an @-pattern being a constant

### LINKED ISSUE #27033: Missing shadowing diagnostic with "@" in pattern-match
```rust
fn main() {
  match Some(1) { None @ _ => { } };
  //let None = 1;
}
```
```
<anon>:2:19: 2:27 warning: unused variable: `None`, #[warn(unused_variables)] on by default
<anon>:2   match Some(1) { None @ _ => { } };
                           ^~~~~~~~
```
This is accepted, but probably shouldn't be.  (A normal declaration of a variable named None is rejected.)

### LINKED ISSUE #27077: Incorrect codegen with shadowed constants in a match arm
From @eefriedman's comment in #27033:

```rust
fn main() {
    const C: u8 = 3;
    match 1 {
        C @ 2 => { println!("{}", C); }
        _ => {}
    };
}
```
```
<anon>:4:9: 4:14 warning: unused variable: `C`, #[warn(unused_variables)] on by default
<anon>:4         C @ 2 => { println!("{}", C); }
                 ^~~~~
Switch constants must all be same type as switch value!
  switch i32 %3, label %match_else [
    i8 3, label %match_case
  ]
LLVM ERROR: Broken function found, compilation aborted!
playpen: application terminated with error code 1
```

Not sure how this is happening and the code should be rejected as in #27033, but it does indicate that something in the compiler is probably broken.

cc @rust-lang/compiler

### PR DESCRIPTION
**Pull request by arielb1**
Fixes #27033
Fixes #27077 

r? @alexcrichton

### CODE REVIEW

**Comment by alexcrichton** [comment]
On the surface it looks like this could be a breaking change, but do the cases this allows basically just always have a codegen error? In that case it seems fine to not worry about running crater.

**Comment by arielb1** [comment]
@alexcrichton 

Actually thinking about it, if you have a unit struct you can do `S @ S` or something and it should work.

**Comment by alexcrichton** [comment]
Hm ok, I'll run a crater run just to be safe, may be good to see if this shows up in the wild much.

File: src/test/compile-fail/issue-27033.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn main() {
+    match Some(1) {
+        None @ _ => {} //~ ERROR declaration of `None` shadows an enum variant
+    };
+    const C: u8 = 1;
+    match 1 {
+        C @ 2 => { //~ ERROR only irrefutable patterns allowed here
```
Comment (eefriedman) [inline src/test/compile-fail/issue-27033.rs]: This error message is really terrible.

Comment (arielb1) [inline src/test/compile-fail/issue-27033.rs]: That's the error you get when you `let C = 0;`

**Comment by alexcrichton** [comment]
Crater says [zero regressions](https://gist.github.com/anonymous/cef5e4383787e2c5b67f), so I'm going to r+.

@bors: r+ 757b0c176fa185fda59283efaf0d8079b702ca69

**Pull request merged**